### PR TITLE
Implementing retries on git commands to work around GHA 128-error issues

### PR DESCRIPTION
**Pull request by ghost**
This PR adds a retry around git commands because `git reset` and `git checkout` regularly fails with a 128 exit code without any good reason.

### CODE REVIEW

File: vcs/github/github.go [hunk 1]
```diff
@@ -290,25 +291,100 @@ func (w *workingCopy) Close() error {
...
+	tries := 0
+	const maxTries = 10
+	for {
+		err := w.g.git(ctx, w.dir, nil, "checkout", string(version))
+		if err == nil {
+			w.version = version
+			return nil
+		}
 		// Checkout failed, see if tag exists.
 		tagExists, e := w.tagExists(ctx, version)
```
Comment (cam72cam) [inline vcs/github/github.go:305]: Do you want to fall back to a tag, even on a 128 error?

**Comment by cam72cam** [comment]
This amount of copy-paste with sensitive code gives me the heebie-jeebies.  I think I'd like to instead see a function that encapsulates most of this.

```go 
func git_with_retry(ctx context.Context, dir string, stdout io.Writer, params ...string) error {

	// We are implementing a retry here because git seems to have issues on GitHub Actions, possibly due to a race
	// condition or inadequate cleanup.
	tries := 0
	const maxTries = 10
	
           for {
		err := w.g.git(ctx, dir, stdout, params)
		if err == nil {
			return nil
		}
		var exitErr *exec.ExitError
		if !errors.As(err, &exitErr) {
			return err
		}
		if exitErr.ExitCode() != 128 {
			return err
		}
		tries += 1
		if tries > maxTries {
			return customThingsAreReallyBrokenError("git %s exited with a 128 error code %d times in a row, aborting (%w)", params[0], maxTries, err)
		}
		select {
		case <-ctx.Done():
			return err
		case <-time.After(100 * time.Millisecond):
			w.g.config.Logger.Debugf(ctx, "git %s exited with a 128 error code, retrying...", params[0])
		}
	}
	```

File: vcs/github/github.go [hunk 2]
```diff
@@ -299,16 +304,90 @@ func (w *workingCopy) checkout(ctx context.Context, version vcs.VersionNumber) e
...
+			return handleVersionNotFound(err)
+		}
+		tries += 1
+		if tries > maxTries {
+			return fmt.Errorf("git checkout exited with a 128 error code %d times in a row, aborting (%w)", maxTries, err)
+		}
+		select {
+		case <-ctx.Done():
+			return err
+		case <-time.After(100 * time.Millisecond):
```
Comment (ollevche) [inline vcs/github/github.go:327]: nit: time.After could negatively affect memory consumption in go versions before 1.23

https://pkg.go.dev/time#After

File: vcs/github/github.go [hunk 3]
```diff
@@ -299,16 +304,90 @@ func (w *workingCopy) checkout(ctx context.Context, version vcs.VersionNumber) e
...
+	}
 }
 
 func (w *workingCopy) clean(ctx context.Context) error {
-	return w.g.git(ctx, w.dir, nil, "clean", "-fd")
+	// We are implementing a retry here because git seems to have issues on GitHub Actions, possibly due to a race
+	// condition or inadequate cleanup.
+	tries := 0
+	const maxTries = 10
+	for {
```
Comment (ollevche) [inline vcs/github/github.go:368]: This retry logic could be separated and reused. I would prefer it that way to ensure all the retries behave the same way for different commands.

File: vcs/github/github.go [hunk 4]
```diff
@@ -398,7 +508,12 @@ func (g github) git(ctx context.Context, dir string, stdout io.Writer, params ..
 	select {
 	case <-done:
 	case <-ctx.Done():
-		_ = cmd.Process.Kill()
+		_ = cmd.Process.Signal(syscall.SIGTERM)
+		select {
+		case <-time.After(30 * time.Second):
+			_ = cmd.Process.Kill()
+		case <-done:
+		}
```
Comment (ollevche) [inline vcs/github/github.go:516]: Could you please explain this part a bit? Not sure I understand it correctly

File: internal/retry/func.go [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+	ctx context.Context,
+	description string,
+	what func() error,
+	isRetryable func(err error) bool,
+	maxTries int,
+	waitTime time.Duration,
+	logger logger.Logger,
+) error {
+	tries := 0
+	for {
```
Comment (cam72cam) [inline internal/retry/func.go:24]: Suggestion, use Func2 with no result to reduce copy paste
```
       _, err := Func2(ctx, description, func() (interface{}, error) {
               return nil, what()
       }, isRetryable, maxTries, waitTime, logger)
       return err
```

**Review by cam72cam — APPROVED** [review]

**Pull request merged**